File: mcp/pmll_memory_mcp/embeddings.py
"""
embeddings.py — Fixed-dimension hashing vectorizer for PMLL memory graph.

Designed to improve context retention and retrieval for coding agents.
Uses the hashing trick so embeddings are stable: querying or embedding new
documents does NOT mutate a global vocabulary. Vectors are always
``EMBED_DIM``-dimensional and L2-normalized.

The legacy ``TfIdfVectorizer`` remains for experiments/export, but ``embed()``
uses the hashing path and never calls ``add_document``.
"""

from __future__ import annotations

import hashlib
import math
import re
import threading
from typing import Dict, List, Optional, Set

EMBED_DIM = 128
_store_lock = threading.RLock()


def tokenize(text: str) -> List[str]:
    """Normalize and tokenize text into lowercase word tokens."""
    text = text.lower()
    text = re.sub(r"[^a-z0-9\s_-]", " ", text)
    return [t for t in text.split() if len(t) > 1]


def _stable_hash(token: str) -> int:
    digest = hashlib.blake2b(token.encode("utf-8"), digest_size=8).digest()
    return int.from_bytes(digest, "little", signed=False)


class HashingVectorizer:
    """Fixed-dimension feature-hashing vectorizer (no mutable vocabulary)."""

    def __init__(self, dim: int = EMBED_DIM) -> None:
        if dim <= 0:
            raise ValueError("dim must be positive")
        self.dim = dim

    @property
    def vocab_size(self) -> int:
        return self.dim

    def vectorize(self, text: str) -> List[float]:
        tokens = tokenize(text)
        vec = [0.0] * self.dim
        if not tokens:
            return vec
        tf: Dict[str, int] = {}
        for token in tokens:
            tf[token] = tf.get(token, 0) + 1
        max_tf = max(tf.values())
        for term, count in tf.items():
            h = _stable_hash(term)
            idx = h % self.dim
            sign = 1.0 if ((h >> 1) & 1) == 0 else -1.0
            normalized_tf = 0.5 + 0.5 * (count / max_tf)
            vec[idx] += sign * normalized_tf
        return l2_normalize(vec)


class TfIdfVectorizer:
    """Legacy TF-IDF vectorizer (mutable vocab — do not use for retrieval)."""

    def __init__(self) -> None:
        self._vocab: Dict[str, int] = {}
        self._idf: Dict[str, float] = {}
        self._doc_count: int = 0
        self._doc_freq: Dict[str, int] = {}

    @property
    def vocab_size(self) -> int:
        return len(self._vocab)

    def add_document(self, text: str) -> None:
        tokens = tokenize(text)
        seen: Set[str] = set()
        for token in tokens:
            if token not in self._vocab:
                self._vocab[token] = len(self._vocab)
            if token not in seen:
                seen.add(token)
                self._doc_freq[token] = self._doc_freq.get(token, 0) + 1
        self._doc_count += 1
        self._recompute_idf()

    def vectorize(self, text: str) -> List[float]:
        tokens = tokenize(text)
        dim = len(self._vocab)
        if dim == 0:
            return []
        tf: Dict[str, int] = {}
        for token in tokens:
            tf[token] = tf.get(token, 0) + 1
        vec = [0.0] * dim
        max_tf = max(tf.values()) if tf else 1
        for term, count in tf.items():
            idx = self._vocab.get(term)
            if idx is not None:
                normalized_tf = 0.5 + 0.5 * (count / max_tf)
                idf = self._idf.get(term, 1.0)
                vec[idx] = normalized_tf * idf
        return l2_normalize(vec)

    def _recompute_idf(self) -> None:
        for term, df in self._doc_freq.items():
            self._idf[term] = math.log(1 + self._doc_count / (1 + df))


def l2_normalize(vec: List[float]) -> List[float]:
    norm = math.sqrt(sum(v * v for v in vec))
    if norm > 1e-10:
        return [v / norm for v in vec]
    return list(vec)


def cosine_similarity(a: List[float], b: List[float]) -> float:
    """Cosine similarity; pads the shorter vector with zeros so dims align."""
    if not a or not b:
        return 0.0
    n = max(len(a), len(b))
    dot = norm_a = norm_b = 0.0
    for i in range(n):
        av = a[i] if i < len(a) else 0.0
        bv = b[i] if i < len(b) else 0.0
        dot += av * bv
        norm_a += av * av
        norm_b += bv * bv
    denom = math.sqrt(norm_a) * math.sqrt(norm_b)
    return dot / denom if denom > 0 else 0.0


_global_hasher: Optional[HashingVectorizer] = None
_global_vectorizer: Optional[TfIdfVectorizer] = None


def get_hasher() -> HashingVectorizer:
    global _global_hasher
    with _store_lock:
        if _global_hasher is None:
            _global_hasher = HashingVectorizer(EMBED_DIM)
        return _global_hasher


def get_vectorizer() -> TfIdfVectorizer:
    """Legacy accessor — prefer ``get_hasher()`` / ``embed()`` for retrieval."""
    global _global_vectorizer
    with _store_lock:
        if _global_vectorizer is None:
            _global_vectorizer = TfIdfVectorizer()
        return _global_vectorizer


def reset_vectorizer() -> None:
    global _global_hasher, _global_vectorizer
    with _store_lock:
        _global_hasher = None
        _global_vectorizer = None


def embed(text: str) -> List[float]:
    """Stable fixed-dim embedding. Does **not** mutate any global vocabulary."""
    return get_hasher().vectorize(text)

File: mcp/src/embeddings.ts
/**
 * embeddings.ts — Fixed-dimension hashing vectorizer for PMLL memory graph.
 *
 * Designed to improve context retention and retrieval for coding agents.
 * `embed()` is stable: it does NOT mutate a global TF-IDF vocabulary.
 */

export const EMBED_DIM = 128;

export function tokenize(text: string): string[] {
  return text
    .toLowerCase()
    .replace(/[^a-z0-9\s_-]/g, " ")
    .split(/\s+/)
    .filter((t) => t.length > 1);
}

function stableHash(token: string): number {
  // FNV-1a 32-bit
  let h = 0x811c9dc5;
  for (let i = 0; i < token.length; i++) {
    h ^= token.charCodeAt(i);
    h = Math.imul(h, 0x01000193);
  }
  return h >>> 0;
}

export function l2Normalize(vec: number[]): number[] {
  let norm = 0;
  for (const v of vec) norm += v * v;
  norm = Math.sqrt(norm);
  if (norm > 1e-10) {
    return vec.map((v) => v / norm);
  }
  return vec.slice();
}

export function cosineSimilarity(a: number[], b: number[]): number {
  if (a.length === 0 || b.length === 0) return 0;
  const n = Math.max(a.length, b.length);
  let dot = 0, normA = 0, normB = 0;
  for (let i = 0; i < n; i++) {
    const av = i < a.length ? a[i] : 0;
    const bv = i < b.length ? b[i] : 0;
    dot += av * bv;
    normA += av * av;
    normB += bv * bv;
  }
  const denom = Math.sqrt(normA) * Math.sqrt(normB);
  return denom === 0 ? 0 : dot / denom;
}

/** Legacy TF-IDF (mutable vocab) — do not use for retrieval. */
export class TfIdfVectorizer {
  private _vocab: Map<string, number> = new Map();
  private _idf: Map<string, number> = new Map();
  private _docCount = 0;
  private _docFreq: Map<string, number> = new Map();
  get vocabSize(): number { return this._vocab.size; }
  addDocument(text: string): void {
    const tokens = tokenize(text);
    const seen = new Set<string>();
    for (const token of tokens) {
      if (!this._vocab.has(token)) this._vocab.set(token, this._vocab.size);
      if (!seen.has(token)) {
        seen.add(token);
        this._docFreq.set(token, (this._docFreq.get(token) ?? 0) + 1);
      }
    }
    this._docCount++;
    for (const [term, df] of this._docFreq) {
      this._idf.set(term, Math.log(1 + this._docCount / (1 + df)));
    }
  }
  vectorize(text: string): number[] {
    const tokens = tokenize(text);
    const dim = this._vocab.size;
    if (dim === 0) return [];
    const tf = new Map<string, number>();
    for (const token of tokens) tf.set(token, (tf.get(token) ?? 0) + 1);
    const vec = new Array(dim).fill(0);
    const maxTf = Math.max(1, ...tf.values());
    for (const [term, count] of tf) {
      const idx = this._vocab.get(term);
      if (idx !== undefined) {
        const normalizedTf = 0.5 + 0.5 * (count / maxTf);
        vec[idx] = normalizedTf * (this._idf.get(term) ?? 1);
      }
    }
    return l2Normalize(vec);
  }
}

export class HashingVectorizer {
  constructor(public readonly dim: number = EMBED_DIM) {
    if (dim <= 0) throw new Error("dim must be positive");
  }
  get vocabSize(): number { return this.dim; }
  vectorize(text: string): number[] {
    const tokens = tokenize(text);
    const vec = new Array(this.dim).fill(0);
    if (tokens.length === 0) return vec;
    const tf = new Map<string, number>();
    for (const token of tokens) tf.set(token, (tf.get(token) ?? 0) + 1);
    const maxTf = Math.max(...tf.values());
    for (const [term, count] of tf) {
      const h = stableHash(term);
      const idx = h % this.dim;
      const sign = ((h >>> 1) & 1) === 0 ? 1 : -1;
      const normalizedTf = 0.5 + 0.5 * (count / maxTf);
      vec[idx] += sign * normalizedTf;
    }
    return l2Normalize(vec);
  }
}

let _globalHasher: HashingVectorizer | null = null;
let _globalVectorizer: TfIdfVectorizer | null = null;

export function getHasher(): HashingVectorizer {
  if (!_globalHasher) _globalHasher = new HashingVectorizer(EMBED_DIM);
  return _globalHasher;
}

export function getVectorizer(): TfIdfVectorizer {
  if (!_globalVectorizer) _globalVectorizer = new TfIdfVectorizer();
  return _globalVectorizer;
}

export function resetVectorizer(): void {
  _globalHasher = null;
  _globalVectorizer = null;
}

/** Stable fixed-dim embedding — does not mutate any global vocabulary. */
export function embed(text: string): number[] {
  return getHasher().vectorize(text);
}
